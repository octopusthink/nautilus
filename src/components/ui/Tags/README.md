A Tag is a keyword used to label, organise, or categorise a piece of data in a visually distinctive and compact way.

```jsx
import { Heading, Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Heading>Wonder Women</Heading>
  <Paragraph>Once upon a time there were a whole group of lady-types who lived in a beautiful paradise-land called ... well, I forget but it was really pretty and just chock full of badass ladies.</Paragraph>
  <Tags>
    <Tags.Tag color="hotpink">Amazon</Tags.Tag>
    <Tags.Tag color="hotpink">Superhero</Tags.Tag>
    <Tags.Tag color="hotpink">Ladies who lunch</Tags.Tag>
  </Tags>
</React.Fragment>
```

## Usage

Use a Tag to show concise metadata about a given piece of content. Content can have many tags, one tag, or zero tags.

### Use this component for...

- Displaying supplementary data used to categorise a piece of content.
- Categories, tags, dates, statuses, counts, authors, and any other metadata that describes or classifies your content.
- Visually differentiating between multiple categories to which content belongs.
- Filtering, sorting, or organising by categories.

### Don't use this component for...

- Creating a standalone link to a new piece of content. Use a `<Button>` instead.

## Appearance

### Colour

By default, tags appear without any colour.

```jsx
import { Heading } from '@octopusthink/nautilus';

<React.Fragment>
  <Tags><Tags.Tag>12 September 2019</Tags.Tag></Tags>
  <Heading>Breaking news: kittens are fluffy</Heading>
</React.Fragment>
```

Use colour as a cue to differentiate between different types of data. Only pass colour to a tag when it provides additional semantic meaning. Select a colour that helps communicate the content of the label, and ensure itâ€™s used consistently.

```jsx
<Tags>
  <Tags.Tag color="hotpink">Hot pink</Tags.Tag>
  <Tags.Tag color="#cd2f83">Pink</Tags.Tag>
  <Tags.Tag color="navy">Navy</Tags.Tag>
  <Tags.Tag color="aqua">Aqua</Tags.Tag>
  <Tags.Tag color="#053e6c">Blue</Tags.Tag>
  <Tags.Tag color="#8a0000">Red</Tags.Tag>
  <Tags.Tag color="#eeb600">Yellow</Tags.Tag>
</Tags>
```

Nautilus will automatically calculate colour contrast for you and change the colour of text to ensure it has sufficient contrast against the background.

### Numbers

If you're showing a count, tally, or other number in a tag, you might want to use slightly different styling by using the `badge` prop. This styling is best used with numbers of three or fewer digits.

```jsx
<Tags>
  <Tags.Tag color="#3b3f45" badge>42</Tags.Tag>
  <Tags.Tag color="#053e6c" badge>1.3</Tags.Tag>
  <Tags.Tag color="#8a0000" badge>Â£9</Tags.Tag>
  <Tags.Tag color="#eeb600" badge>130</Tags.Tag>
  <Tags.Tag color="#eeb600" badge>1.30%</Tags.Tag>
</Tags>
```

### Semantic status indicators

Tags are often used as a way of indicating the state or status of a process. To facilitate this usage, Nautilus provides a set of semantic, colour-coded status indicators.

```jsx
<Tags>
  <Tags.Tag status="neutral">To-do</Tags.Tag>
  <Tags.Tag status="new">Beta</Tags.Tag>
</Tags>

<Tags>
  <Tags.Tag status="inProgress">In Progress</Tags.Tag>
  <Tags.Tag status="success">Done</Tags.Tag>
</Tags>

<Tags>
  <Tags.Tag status="warning">Warning</Tags.Tag>
  <Tags.Tag status="danger">Failed</Tags.Tag>
</Tags>
```

| Status       | Description               | Usage                                    |
|--------------|---------------------------|------------------------------------------|
| `neutral`    | Unspecified or unstarted. | To-do, unavailable, minor, not started.
| `new`        | New, updated, or beta.    | New, trial, information, help, alpha, beta.
| `inProgress` | In progress or current.   | In progress, open, modified.
| `success`    | Completed or successful.  | Available, done, approved, resolved, added, complete.
| `danger`     | Failure or deletion.      | Errors, declined, failed, removed, critical.
| `warning`    | Attention or hazard.      | Busy, blocked, missing, warning, information.

Colour mappings can be customised in your theme.

### Theming

Generally speaking, the styling of Tags should be controlled in your theme. Tags should look the same across your site so that users can immediately recognise their purpose.

Capitalisation, colour placement (background or border), border-radius, and padding are all customisable via your theme.

## Behaviour

### A tag can have an associated action

A Tag can act as a link or a button, depending on context. Clicking on the tag should direct the user to more information about that tag. For instance, clicking the tag may take the user to a page listing only posts with that tag, or it may filter the current list to show/hide content using the tag.

### A tag can be removed

You may want to use Tags to allow users to add and remove attributes from an object. Do this by passing an event function to the Tag's `onDismiss` prop. This event will be called when the Tag's "dismiss" button is activated.

```jsx
const onDismiss = (event) => {
  window.alert('Goodbye Tag! ðŸ‘‹');
};

<Tags>
  <Tags.Tag color="#cd2f83" onDismiss={onDismiss}>Dismiss me!</Tags.Tag>
</Tags>
```
When a Tag is removable, it will be indicated with a "dismiss" button (stylised as an `X`/"close" icon).

### Placement

Tags should be in close proximity to the data that they're labelling so as to clearly identify and reinforce the relationship.

## Voice & tone

- Tag labels should use sentence case.
- Tag labels should be short, succinct, and easily scannable. Avoid tags that are longer than three words.
- If you need more than one word in a tag label, separate the words with a space.

## Accessibility

### Typography and type case

Nautilus defaults to showing tag labels in small-caps to distinguish them from body copy and interactive elements.

Small caps can be difficult for users to parse, so it's important to use them sparingly. Remember that Tags are intended only to present *supplementary* data that isn't critical to understanding the core content. Tags should *never* contain content or data that's essential to understanding the content you're presenting.

Use sentence case when passing text strings to a Tag. Small caps styling will be added by CSS. If you pass a word in all caps, screen readers can behave unpredictably. **Only** pass labels in all caps if they're abbreviations and should be read out loud as such.

### Grouping Tags

When showing a group or list of tags, use a Group or List component to indicate that tags are part of the same grouping.

Keep in mind that tags increase the amount of visual noise, especially when using multiple colours, so use them in moderation. Only use tags that provide helpful additional data to users.

### Label text

Tags should include a label text to indicate what type of data is being presented. This ensures that users of assistive technology can understand the context of the Tag, even if they may not be able to see the Tag styling.

To do this, include a `label` attribute on the Tag. This label will be read by screen readers, but will be hidden from the visual display.

```jsx
import { Heading } from '@octopusthink/nautilus';

<React.Fragment>
  <Tags><Tags.Tag label="Category">Books</Tags.Tag></Tags>
  <Heading>Little Women</Heading>
  <Tags><Tags.Tag label="Date">10 November</Tags.Tag></Tags>
  <Heading>The day the music died</Heading>
</React.Fragment>
```

If no `label` property is passed, the Tag will be prefaced with a "Tag" label. Status Tags automatically include this extra text.

### ARIA properties

The button to remove a Tag is given an `aria-label` so that screen reader users can distinguish which Tag will be removed.
