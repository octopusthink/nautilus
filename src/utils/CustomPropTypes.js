import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
