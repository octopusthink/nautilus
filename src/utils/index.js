export { default as CustomPropTypes } from './CustomPropTypes';
