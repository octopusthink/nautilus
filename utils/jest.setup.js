// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
