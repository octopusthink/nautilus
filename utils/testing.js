// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
