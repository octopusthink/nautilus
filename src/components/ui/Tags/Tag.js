import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { Icon } from 'components/ui/Icon';
import { getContrastingTextColor, metadata, toUnits } from 'styles';
import { useTheme } from 'themes';

export const Tag = (props) => {
  const [isDismissed, setDismissed] = useState(false);
  const theme = useTheme();
  const { badge, children, color, onDismiss, status, ...otherProps } = props;

  if (isDismissed) {
    return null;
  }

  let textColor = theme.colors.text.light;
  let backgroundColor;
  if (color) {
    backgroundColor = color;
  }

  if (status) {
    backgroundColor = theme.colors.intent[status];
  }

  if (status || color) {
    textColor = getContrastingTextColor({ color: backgroundColor, theme });
  }

  return (
    <span
      css={css`
        ${metadata.small(theme)};
        color: ${textColor};
        margin: 0 ${toUnits(theme.spacing.padding.xSmall)} ${toUnits(theme.spacing.padding.xSmall)}
          0;

        ${backgroundColor &&
          css`
            background: ${backgroundColor};
            padding: ${toUnits(theme.spacing.padding.xSmall)}
              ${toUnits(theme.spacing.padding.small)};
          `}

        ${badge &&
          css`
            border-radius: ${toUnits(theme.spacing.padding.xLarge)};
            min-width: ${toUnits(theme.spacing.padding.xLarge)};
            height: ${toUnits(theme.spacing.padding.xLarge)};
            display: inline-flex;
            align-items: center;
            justify-content: center;
            box-sizing: border-box;
          `}
      `}
      {...otherProps}
    >
      {children}
      {onDismiss && (
        <button
          css={css`
            box-shadow: 0;
            border: 0;
            background: none;
            color: currentColor;
            cursor: pointer;
            vertical-align: 0.2rem;
            margin: 0;
            padding: 0;

            &:hover {
              opacity: 0.75;
            }
          `}
          onClick={(event) => {
            event.preventDefault();

            setDismissed(true);

            onDismiss(event);
          }}
          type="button"
        >
          <Icon
            name="x"
            small
            css={css`
              display: inline-block;
              margin-left: ${toUnits(theme.spacing.padding.small)};
              margin-top: ${toUnits(theme.spacing.padding.xxSmall)};
              opacity: 1;
              vertical-align: middle;
            `}
          />
        </button>
      )}
    </span>
  );
};

Tag.defaultProps = {
  children: undefined,
  color: null,
  badge: false,
  onDismiss: null,
  status: undefined,
};

Tag.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Use colour to differentiate different tags or indicate status. */
  color: PropTypes.string,
  /** Style numbers or counts with a badge. */
  badge: PropTypes.bool,
  /** Function to call when a Tag is dismissed via the close button. */
  onDismiss: PropTypes.func,
  /** Indicate status using a semantic colour set. */
  status: PropTypes.oneOf(['neutral', 'success', 'warning', 'danger', 'inProgress', 'new']),
};

export const { defaultProps, propTypes } = Tag;

export default Tag;
