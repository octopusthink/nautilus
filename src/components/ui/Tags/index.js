import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React, { Children, cloneElement, Fragment, useMemo, useState } from 'react';
import shortid from 'shortid';

import List from 'components/ui/List';
import Paragraph from 'components/ui/Paragraph';
import { ComponentClassName as ListItemClassName } from 'components/ui/List/Item';

import Tag from './Tag';

export const Tags = (props) => {
  const { children, label } = props;

  const [generatedId] = useState(shortid.generate());

  const items = useMemo(() => {
    const tagElements = Children.toArray(children).filter((child) => {
      return child.type === Tag;
    });

    if (tagElements.length > 1) {
      return (
        // TODO: Export an `UnstyledList` component we can use to prevent
        // override styles like these.
        <List
          aria-labelledby={label && generatedId}
          css={css`
            display: inline-flex;
            list-style: none;
            margin: 0;
            width: 100%;

            > .${ListItemClassName}::before {
              content: '';
            }
          `}
        >
          {tagElements.map((tag, index) => {
            // eslint-disable-next-line react/no-array-index-key
            return <List.Item key={`tag-${index}`}>{tag}</List.Item>;
          })}
        </List>
      );
    }

    if (label) {
      return cloneElement(tagElements[0], { 'aria-labelledby': generatedId });
    }

    return tagElements[0];
  }, [children, generatedId, label]);

  const WrapperComponent = items.length > 1 ? List : Fragment;

  return (
    <>
      {label && (
        <Paragraph // TODO: switch with VisuallyHidden component, once available.
          css={css`
            font-style: bold;
            position: absolute;
            left: -10000px;
            top: auto;
            width: 1px;
            height: 1px;
            overflow: hidden;
          `}
          id={generatedId}
        >
          {label}
        </Paragraph>
      )}
      <WrapperComponent>{items}</WrapperComponent>
    </>
  );
};

Tags.defaultProps = {
  children: undefined,
  label: undefined,
};

Tags.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** A description of the group of tags. */
  label: PropTypes.string,
};

// Export Tag as `Tags.Tag`.
Tags.Tag = Tag;

export default Tags;
