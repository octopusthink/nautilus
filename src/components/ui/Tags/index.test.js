import { css } from '@emotion/react';
import React from 'react';

import { axe, render } from '../../../../utils/testing';
import Tags from '.';

describe('Tags', () => {
  it('should render a <span> tag if only one tag is created with no title prop', () => {
    const { container } = render(
      <Tags>
        <Tags.Tag>Hello</Tags.Tag>
      </Tags>,
    );

    expect(container.firstChild.firstChild.tagName).toEqual('SPAN');
  });

  it('should be styled by a css prop', () => {
    const { container } = render(
      <Tags
        css={css`
          background: hotpink;
        `}
      >
        <Tags.Tag>Hello</Tags.Tag>
      </Tags>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should output only Tags.Tag children', () => {
    const { getByTestId } = render(
      <Tags>
        <Tags.Tag data-testid="child-one">Hello</Tags.Tag>
        <span data-testid="child-two" />
      </Tags>,
    );

    expect(getByTestId('child-one')).toBeDefined();
    expect(() => {
      getByTestId('child-two');
    }).toThrow('Unable to find an element');
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Tags.Tag className="custom-class" data-testid="myText">
        hello
      </Tags.Tag>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });

  it('should match Tags styles', () => {
    const { container } = render(
      <Tags>
        <Tags.Tag data-testid="child-one">Hello</Tags.Tag>
        <Tags.Tag data-testid="child-two">Goodbye</Tags.Tag>
      </Tags>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match styles', () => {
    const { container } = render(<Tags.Tag>hello</Tags.Tag>);

    expect(container).toMatchSnapshot();
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(
        <Tags>
          <Tags.Tag data-testid="child-one">Hello</Tags.Tag>
        </Tags>,
      );

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });

    it('should pass aXe tests with a title prop', async () => {
      const { container } = render(
        <Tags label="My tags">
          <Tags.Tag data-testid="child-one">Hello</Tags.Tag>
          <Tags.Tag data-testid="child-two">Goodbye</Tags.Tag>
        </Tags>,
      );

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
