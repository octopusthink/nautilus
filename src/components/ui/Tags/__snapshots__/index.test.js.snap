// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags should match Tags styles 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 700;
  font-size: 1.5rem;
  line-height: 1.3333333333333333;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  text-transform: uppercase;
  color: #666c76;
  margin: 0 0.4rem 0.4rem 0;
}

.emotion-1 {
  position: relative;
  margin: 0 0 0.4rem;
}

.emotion-1::before {
  position: absolute;
  right: 101%;
}

<div>
  <ul
    class="Nautilus-List emotion-4"
  >
    <li
      class="Nautilus-ListItem emotion-1"
    >
      <span
        class="emotion-0"
        data-testid="child-one"
      >
        Hello
      </span>
    </li>
    <li
      class="Nautilus-ListItem emotion-1"
    >
      <span
        class="emotion-0"
        data-testid="child-two"
      >
        Goodbye
      </span>
    </li>
  </ul>
</div>
`;

exports[`Tags should match styles 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 700;
  font-size: 1.5rem;
  line-height: 1.3333333333333333;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  text-transform: uppercase;
  color: #666c76;
  margin: 0 0.4rem 0.4rem 0;
}

<div>
  <span
    class="emotion-0"
  >
    hello
  </span>
</div>
`;
